homeassistant/components/velbus.py

```diff
@@ -26,18 +27,60 @@
 }, extra=vol.ALLOW_EXTRA)
 
 
-def setup(hass, config):
+async def async_setup(hass, config):
     """Set up the Velbus platform."""
     import velbus
     port = config[DOMAIN].get(CONF_PORT)
-    connection = velbus.VelbusUSBConnection(port)
-    controller = velbus.Controller(connection)
+    controller = velbus.Controller(port)
+
     hass.data[DOMAIN] = controller
 
     def stop_velbus(event):
         """Disconnect from serial port."""
         _LOGGER.debug("Shutting down ")
-        connection.stop()
+        controller.stop()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_velbus)
+
+    def callback():
+        modules = controller.get_modules()
+        discovery_info = {
+            'switch': [],
+            'binary_sensor': []
+        }
+        for module in modules:
+            for channel in range(1, module.number_of_channels() + 1):
+                for category in discovery_info:
+                    if category in module.get_categories(channel):
+                        discovery_info[category].append((
+                            module.get_module_address(),
+                            channel
+                        ))
+        load_platform(hass, 'switch', DOMAIN,
+                      discovery_info['switch'], config)
+        load_platform(hass, 'binary_sensor', DOMAIN,
+                      discovery_info['binary_sensor'], config)
+
+    controller.scan(callback)
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_velbus)
     return True
+
+
+class VelbusEntity:
+    """Representation of a Velbus entity."""
+
+    def __init__(self, module, channel, hass):
+        """Initialize a Velbus entity."""
+        self._module = module
+        self._channel = channel
+        self.hass = hass
```

Comment:
Remove this. `hass` will be set on the entity when it has been added to home assistant.